flink/src/main/java/org/apache/iceberg/flink/source/enumerator/IcebergSplitAssigner.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.flink.source.enumerator;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.Optional;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplit;
+
+public interface IcebergSplitAssigner {
+
+  /**
+   * Gets the next split.
+   *
+   * <p>When this method returns an empty {@code Optional}, then the set of splits is
+   * assumed to be done and the source will finish once the readers finished their current
+   * splits.
+   */
+  Optional<IcebergSourceSplit> getNext(int subttask);
```

Comment:
I think this interface is too restrictive IMO. A split completion can potentially complete trigger assignment of splits to subtasks not necessarily responsible for completing the split. A more flexible and robust method would be to change this to Map<Integer, List<Split>> getGoalState();

Reply:
I agree that current interface won't be enough for the event time alignment assigner. what if we change it to return a `CompleteableFuture`. `SourceCoordinatorContext.assignSplits` ensures `callInCoordinatorThread`.

Here is what I am thinking on the event time alignment assigner can work.

- Upon getNext, it can either complete the future immediately or return an incomplete future. incomplete future is added to a queue internally
- when watermark advances, assigner checks the queue to see if there is any pending futures that are allowed to be completed under the event time alignment constraint. This probably should be called from a separator thread/executor internal to the assigner, as watermark advancement may be initiated by `onSplitsCompleted`.
- assigner can optionally schedule a fixed interval checking on incomplete futures in the queue